02-intermediate-ruby/exception-handling.md

```diff
@@ -0,0 +1,313 @@
+
+rescue ArgumentError
+  # Code to recover from an ArgumentError
+
+rescue ZeroDivisionError
+  # Code to recover from a ZeroDivisionError
+
+rescue
+  # other exceptions handled here
+
+end
+```
+
+If you specify a type of exception, Ruby will rescue that exception as well as any of its subclasses. That means that `rescue Exception` will rescue all exceptions, since all exceptions are subclasses of `Exception`.
+
+Most recoverable errors inherit from `StandardError`, and without any arguments `rescue` will only rescue `StandardError` and its subclasses.
+
+It is almost always worth your time to figure out exactly what type of exception you're expecting, and rescue only that. Otherwise you're setting yourself up to ignore the valuable information Ruby is giving you about why your program failed. A bare `rescue` with no exception type specified is a classic example of sloppy programming.
+
+### Accessing a Rescued Exception
+
+Exceptions are full of useful information about what went wrong, which is often useful inside a `rescue` clause. Ruby allows us to access the thrown exception using `=>`
+
+```ruby
+begin
+	# do something risky
+rescue ArgumentError => exception
+	puts "Encountered an error: #{exception}"
+end
+```
+
+Like method or block parameters, the rescued exception is a local variable and may be called anything you please.
+
+### Example: `csv_printer.rb`
+
+```ruby
+# csv_printer.rb
+require 'csv'
+
+while true
+  puts "What file would you like to print?"
+  filename = gets.chomp
+
+  begin
+    CSV.readlines(filename).each do |line|
+      puts line.join(",")
+    end
+
+  rescue SystemCallError => exception
+    puts "Could not open file: #{exception.message}"
+  end
+
+  puts "Would you like to go again?"
+  break if gets.chomp != "yes"
+end
+```
+
+### In Minitest
+
+If we expect a method to raise an exception and it doesn't, that's a bug! This means that we must test that our methods raise the exceptions we want, as well as testing nominal behavior. As we saw on the BankAccounts project, you can look for an exception in a given block of code in Minitest like this
+
+```ruby
+it "Raises an ArguemntError when asked to withdraw a negative amount" do
+	account = Bank::Account.new(1337, 10000)
+	proc {
+		account.withdraw(-100)
+	}.must_raise ArgumentError
+end
+```
+
+## Creating Custom Exceptions
+
+Sometimes you run into a situation where the built in Ruby exceptions don't quite describe what went wrong. For example, imagine you have a program for processing credit card transactions, and you need an exception for an invalid card number. Lucky for us, Ruby allows you to define custom exceptions!
+
+As we discussed above, exceptions are just classes that inherit from `Exception` or one of its subclasses, so to making your own is as simple as defining a new class. You'll usually want to inherit from `StandardError`, so that your exception can be handled with `rescue`.
+
+```ruby
+class InvalidCardNumberError < StandardError
```

Comment:
I guess this example doesn't quite prove to me why we'd want to create a new exception of our own.